### PR TITLE
fix(rust): Prevent panic on sample_n with replacement from empty df

### PR DESCRIPTION
**Pull request by trueb2**
Sampling n from an empty dataframe panics if with_replacement=True. A test case is added documenting the behavior of sample_n and sample_frac for with_replacement and without.

I am not really sure why ensure_shape is used without symmetry, but it appears intentional.

```rust
let subdf = df.sample_n(50000, true, false, None)?;
```
 

```
shape: (0, 4)
┌──────────────┬─────┬─────┬─────┐
│ t            ┆ x   ┆ y   ┆ z   │
│ ---          ┆ --- ┆ --- ┆ --- │
│ datetime[μs] ┆ f64 ┆ f64 ┆ f64 │
╞══════════════╪═════╪═════╪═════╡
└──────────────┴─────┴─────┴─────┘

thread 'tokio-runtime-worker' panicked at 'Uniform::new called with `low >= high`', /Users/jwtrueb/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rand-0.8.5/src/distributions/uniform.rs:567:1
   0: rust_begin_unwind
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/panicking.rs:593:5
   1: core::panicking::panic_fmt
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/panicking.rs:67:14
   2: core::panicking::panic
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/panicking.rs:117:5
   3: <rand::distributions::uniform::UniformInt<u32> as rand::distributions::uniform::UniformSampler>::new
             at /Users/jwtrueb/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rand-0.8.5/src/distributions/uniform.rs:452:17
   4: rand::distributions::uniform::Uniform<X>::new
             at /Users/jwtrueb/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rand-0.8.5/src/distributions/uniform.rs:189:17
   5: polars_core::chunked_array::random::create_rand_index_with_replacement
             at /Users/jwtrueb/.cargo/registry/src/index.crates.io-6f17d22bba15001f/polars-core-0.32.1/src/chunked_array/random.rs:18:16
   6: polars_core::chunked_array::random::<impl polars_core::frame::DataFrame>::sample_n
             at /Users/jwtrueb/.cargo/registry/src/index.crates.io-6f17d22bba15001f/polars-core-0.32.1/src/chunked_array/random.rs:177:21
   7: qsib_data_viz::sample_data::{{closure}}::{{closure}}
             at ./src/main.rs:306:25
```

### CODE REVIEW

File: crates/polars-core/src/chunked_array/random.rs [hunk 1]
```diff
@@ -9,6 +9,9 @@ use crate::random::get_global_random_u64;
 use crate::utils::{CustomIterTools, NoNull};
 
 fn create_rand_index_with_replacement(n: usize, len: usize, seed: Option<u64>) -> IdxCa {
+    if len == 0 {
+        return NoNull::<IdxCa>::from_iter(std::iter::empty()).into_inner();
```
Comment (ritchie46) [inline crates/polars-core/src/chunked_array/random.rs:13]: Can you use `IdxCa::new_empty`

File: crates/polars-core/src/chunked_array/random.rs [hunk 2]
```diff
@@ -257,6 +260,22 @@ impl BooleanChunked {
 mod test {
     use super::*;
 
+    #[test]
+    fn test_sample_empty_df() {
```
Comment (ritchie46) [inline crates/polars-core/src/chunked_array/random.rs:264]: Can you move this test to python?

**Review by ritchie46 — COMMENTED** [review]
Thanks for the fix. I have left a few minor comments.

**Comment by trueb2** [comment]
I added the test to py-polars and confirmed that without the fix we hit a panic
```
E       pyo3_runtime.PanicException: Uniform::new called with `low >= high`
```

I do not think there is a way to use `IdxCa::new_empty`. IdxCa is a ChunkedArray and the only new_empty is on Series or an ObjectChunked. I am updating to something more concise `IdxCa::new_vec("", vec![]);`

**Comment by trueb2** [comment]
Nice CI and make setup! This was pretty streamlined and intuitive for making contributions for the first time to this project!

**Comment by ritchie46** [comment]
> Nice CI and make setup! This was pretty streamlined and intuitive for making contributions for the first time to this project!

Nice! Honours to @stinodego on this one. ;)

**Pull request merged**